### PR TITLE
initial unit tests for the converter

### PR DESCRIPTION
**Pull request by xeno-by**
Supersedes https://github.com/scalameta/paradise/pull/58

### CODE REVIEW

**Comment by xeno-by** [comment]
Review @iTakeshi @olafurpg

File: plugin/src/main/scala/org/scalameta/paradise/reflect/LogicalTrees.scala [hunk 1]
```diff
@@ -579,9 +576,9 @@ trait LogicalTrees { self: ReflectToolkit =>
     object PackageObjectDef {
       def unapply(tree: g.PackageDef): Option[(List[l.Modifier], l.TermName, l.Template)] = {
         if (!tree.is(PackageObjectPackageRole)) return None
-        val PackageObjectPackageRole(g.ModuleDef(_, _, templ)) = tree.get(PackageObjectPackageRole)
-        val ltempl                                             = l.Template(templ, l.TermName(tree))
-        Some((Nil, ???, ltempl))
+        val PackageObjectPackageRole(module @ g.ModuleDef(_, _, templ)) =
+          tree.get(PackageObjectPackageRole)
+        Some((Nil, ???, l.Template(module)))
```
Comment (olafurpg) [inline plugin/src/main/scala/org/scalameta/paradise/reflect/LogicalTrees.scala]: Why is the `???` required? Even though it was there before, a small comment would be nice.

**Comment by xeno-by** [comment]
The main problem that we need to solve is desugarings as explained by @olafurpg in https://github.com/scalameta/paradise/pull/58#issuecomment-252596133.

Some desugarings are automatically undone by the converter, but some of them, e.g. representing infix applications with normal applications are impossible to undesugar. That's why this PR features a really ugly test: `syntactic("def add(a: Int)(implicit z: Int = 0) = a.+(z)")`.

We have two alternatives, as far as I can see:
  1. Provide both the original and the desugared code snippets when they are different, as proposed by Olaf in the aforementioned comment.
  1. Write a more complicated tree comparer instead of comparing structures. Such comparer could be aware of things like ApplyInfix -> Apply.

Wdyt?

**Comment by xeno-by** [comment]
Also, I'd like to give credit to @liufengyun for his Eden project. Our new test suite is directly inspired by https://github.com/liufengyun/eden/blob/master/src/test/scala/dotty/eden/UntpdSuite.scala.

File: plugin/src/main/scala/org/scalameta/paradise/reflect/LogicalTrees.scala [hunk 2]
```diff
@@ -651,15 +648,19 @@ trait LogicalTrees { self: ReflectToolkit =>
...
-              if scala.symbol == definitions.ScalaPackage =>
-            true
-          case _ => false
+      def apply(tree: g.ImplDef): l.Template = {
+        def removeSyntheticParents(parents: List[g.Tree]): List[g.Tree] = parents match {
+          case List(anyRef) if anyRef.toString == "scala.AnyRef" =>
+            Nil
+          case parents :+ product :+ serializable
+              if tree.mods
+                .hasFlag(CASE) && product.toString == "scala.Product" && serializable.toString == "scala.Serializable" =>
```
Comment (olafurpg) [inline plugin/src/main/scala/org/scalameta/paradise/reflect/LogicalTrees.scala]: FYI, you control line breaks for all infix operators. Scalafmt does not mess with boolean operators and DSLs.

**Comment by liufengyun** [comment]
The idea for testing goes back to @olafurpg  👍 !

**Comment by xeno-by** [comment]
Allright, merging

**Comment by iTakeshi** [comment]
Very excited to see your sophisticated implementation! Thank you for upgrading my primitive one.

On the "desugaring" problem:
I still don't understand @olafurpg 's proposal in https://github.com/scalameta/paradise/pull/58#issuecomment-252596133 well. Do you mean something like that?
```scala
val origianl: String = "some source code"
val desugared: g.Tree = parseReflect(original)
if (original == desugared.toString) {
  test(toMtree(desugared) == parseMeta(original))
} else {
  test(
    toMtree(undesugar(desugared)) == parseMeta(original)
    ||
    toMtree(desugared) == parseMeta(original)
  )
}
```

If this understanding is correct for option 1 in @xeno-by 's comment, I don't like this idea because undesugaring can fail in some edge cases as you mentioned above. Although we might be able to write undesugaring function for such cases, it will be relatively harder.

I support option 2 because it will be nice in long-term perspective to make this test suite robust, even in the edge cases. Also, such comparer can solve this https://github.com/scalameta/paradise/issues/43#issuecomment-250155972
